compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -1351,4 +1352,132 @@ object messages {
+    val kind = "Type Mismatch"
+    val explanation =
+      hl"""To create an instance of $tp it needs to inherit $selfType in some way.
+          |
+          |Note: Self types are indicated with the notation
+          |  ${s"class "}$tp ${"{ this: "}$selfType${" => "}
+          |"""
+  }
+
+  case class AbstractMemberMayNotHaveModifier(sym: Symbol, flag: FlagSet)(
+    implicit ctx: Context)
+    extends Message(AbstractMemberMayNotHaveModifierID) {
+    val msg = hl"""abstract $sym may not have `$flag' modifier"""
+    val kind = "Syntax"
+    val explanation = ""
+  }
+
+  case class TopLevelCantBeImplicit(sym: Symbol)(
+    implicit ctx: Context)
+    extends Message(TopLevelCantBeImplicitID) {
+    val msg = hl"""${"implicit"} modifier cannot be used for top-level definitions"""
+    val kind = "Syntax"
+    val explanation = ""
+  }
+
+  case class TypesAndTraitsCantBeImplicit(sym: Symbol)(
+    implicit ctx: Context)
+    extends Message(TypesAndTraitsCantBeImplicitID) {
+    val msg = hl"""${"implicit"} modifier cannot be used for types or traits"""
+    val kind = "Syntax"
+    val explanation = ""
+  }
+
+  case class OnlyClassesCanBeAbstract(sym: Symbol)(
+    implicit ctx: Context)
+    extends Message(OnlyClassesCanBeAbstractID) {
+    val msg = hl"""${"abstract"} modifier can be used only for classes; it should be omitted for abstract members"""
+    val kind = "Syntax"
+    val explanation = ""
+  }
+
+  case class AbstractOverrideOnlyInTraits(sym: Symbol)(
+    implicit ctx: Context)
+    extends Message(AbstractOverrideOnlyInTraitsID) {
+    val msg = hl"""${"abstract override"} modifier only allowed for members of traits"""
+    val kind = "Syntax"
+    val explanation = ""
+  }
+
+  case class TraitsMayNotBeFinal(sym: Symbol)(
+    implicit ctx: Context)
+    extends Message(TraitsMayNotBeFinalID) {
+    val msg = hl"""$sym may not be ${"final"}"""
+    val kind = "Syntax"
+    val explanation = "A trait can never be final."
+  }
+
+  case class NativeMembersMayNotHaveImplementation(sym: Symbol)(
+    implicit ctx: Context)
+    extends Message(NativeMembersMayNotHaveImplementationID) {
+    val msg = hl"""${"@native"} members may not have an implementation"""
+    val kind = "Syntax"
+    val explanation = ""
+  }
+
+  case class OnlyClassesCanHaveDeclaredButUndefinedMembers(sym: Symbol)(
+    implicit ctx: Context)
+    extends Message(OnlyClassesCanHaveDeclaredButUndefinedMembersID) {
+
+    private val varNote =
+      if (sym.is(Mutable)) "Note that variables need to be initialized to be defined."
+      else ""
+    val msg = hl"""only classes can have declared but undefined members"""
+    val kind = "Syntax"
+    val explanation = s"$varNote"
+  }
+
+  case class CannotExtendAnyVal(sym: Symbol)(implicit ctx: Context)
+    extends Message(CannotExtendAnyValID) {
+    val msg = hl"""$sym cannot extend ${"AnyVal"}"""
+    val kind = "Syntax"
+    val explanation = ""
```

Comment:
This leads to the question how we'd support adding relevant links to `Message`s. Links to the lang specs will be helpful in many cases, Stackoverflow might be good, too.
But links do not always age well.